File: Server/Systems/Paths/CircularPath.cs
using OceanKing.Server.Models;

namespace OceanKing.Server.Systems.Paths;

/// <summary>
/// Circular or elliptical movement pattern
/// </summary>
public class CircularPath : IPath
{
    private readonly float[] _center;
    private readonly float _radiusX;
    private readonly float _radiusY;
    private readonly float _startAngle;
    private readonly bool _clockwise;
    private readonly int _fishId;
    private readonly int _seed;
    private readonly int _startTick;
    private readonly float _speed;
    
    public CircularPath(int fishId, int seed, int startTick, float speed, float[] center, float radiusX, float radiusY, float startAngle, bool clockwise)
    {
        _fishId = fishId;
        _seed = seed;
        _startTick = startTick;
        _speed = speed;
        _center = center;
        _radiusX = radiusX;
        _radiusY = radiusY;
        _startAngle = startAngle;
        _clockwise = clockwise;
    }
    
    public float[] GetPosition(float t)
    {
        // Calculate angle based on progress
        float angle = _startAngle + (t * MathF.PI * 2);
        if (_clockwise)
        {
            angle = _startAngle - (t * MathF.PI * 2);
        }
        
        float x = _center[0] + MathF.Cos(angle) * _radiusX;
        float y = _center[1] + MathF.Sin(angle) * _radiusY;
        
        return new[] { x, y };
    }
    
    public PathType GetPathType() => PathType.Circular;
    
    public PathData GetPathData()
    {
        // Circumference approximation for ellipse: π * (3(a+b) - sqrt((3a+b)(a+3b)))
        float a = _radiusX;
        float b = _radiusY;
        float circumference = MathF.PI * (3 * (a + b) - MathF.Sqrt((3 * a + b) * (a + 3 * b)));
        
        return new PathData
        {
            FishId = _fishId,
            PathType = PathType.Circular,
            Seed = _seed,
            StartTick = _startTick,
            Speed = _speed,
            ControlPoints = new[] 
            { 
                _center, 
                new[] { _radiusX, _radiusY }, 
                new[] { _startAngle, _clockwise ? 1f : 0f }
            },
            Duration = circumference / _speed,
            Loop = true
        };
    }
}

File: src/systems/paths/CircularPath.ts
import { PathData, PathType } from './PathData';

/**
 * Circular or elliptical movement pattern
 * Exact port of C# CircularPath implementation
 */
export class CircularPath {
  private readonly center: [number, number];
  private readonly radiusX: number;
  private readonly radiusY: number;
  private readonly startAngle: number;
  private readonly clockwise: boolean;
  private readonly fishId: number;
  private readonly seed: number;
  private readonly startTick: number;
  private readonly speed: number;

  constructor(
    fishId: number,
    seed: number,
    startTick: number,
    speed: number,
    center: [number, number],
    radiusX: number,
    radiusY: number,
    startAngle: number,
    clockwise: boolean,
  ) {
    this.fishId = fishId;
    this.seed = seed;
    this.startTick = startTick;
    this.speed = speed;
    this.center = center;
    this.radiusX = radiusX;
    this.radiusY = radiusY;
    this.startAngle = startAngle;
    this.clockwise = clockwise;
  }

  /**
   * Get position along the circular/elliptical path at normalized time t
   * Matches C# CircularPath.GetPosition (lines 32-43)
   */
  getPosition(t: number): [number, number] {
    let angle = this.startAngle + t * Math.PI * 2;
    if (this.clockwise) {
      angle = this.startAngle - t * Math.PI * 2;
    }

    const x = this.center[0] + Math.cos(angle) * this.radiusX;
    const y = this.center[1] + Math.sin(angle) * this.radiusY;

    return [x, y];
  }

  getPathType(): PathType {
    return PathType.Circular;
  }

  /**
   * Get serializable path data
   * Matches C# CircularPath.GetPathData (lines 47-67)
   */
  getPathData(): PathData {
    const a = this.radiusX;
    const b = this.radiusY;
    const circumference = Math.PI * (3 * (a + b) - Math.sqrt((3 * a + b) * (a + 3 * b)));

    return {
      fishId: this.fishId,
      pathType: PathType.Circular,
      seed: this.seed,
      startTick: this.startTick,
      speed: this.speed,
      controlPoints: [
        this.center,
        [this.radiusX, this.radiusY],
        [this.startAngle, this.clockwise ? 1 : 0],
      ],
      duration: circumference / this.speed,
      loop: true,
    };
  }
}
